hathor/p2p/node_sync_v2.py

```diff
@@ -0,0 +1,1188 @@
+        return self.protocol.peer.id[:7]
+
+    @property
+    def is_started(self) -> bool:
+        return self._started
+
+    def start(self) -> None:
+        """ Start sync.
+        """
+        if self._started:
+            raise Exception('NodeSyncBlock is already running')
+        self._started = True
+        self.reactor.callLater(0, self.run_sync)
+
+    def stop(self) -> None:
+        """ Stop sync.
+        """
+        if not self._started:
+            raise Exception('NodeSyncBlock is already stopped')
+        self._started = False
+        if self.call_later_id and self.call_later_id.active():
+            self.call_later_id.cancel()
+
+    def get_cmd_dict(self) -> Dict[ProtocolMessages, Callable[[str], None]]:
+        """ Return a dict of messages of the plugin.
+        """
+        return {
+            ProtocolMessages.GET_NEXT_BLOCKS: self.handle_get_next_blocks,
+            ProtocolMessages.GET_PREV_BLOCKS: self.handle_get_prev_blocks,
+            ProtocolMessages.BLOCKS: self.handle_blocks,
+            ProtocolMessages.BLOCKS_END: self.handle_blocks_end,
+            ProtocolMessages.GET_BEST_BLOCK: self.handle_get_best_block,
+            ProtocolMessages.BEST_BLOCK: self.handle_best_block,
+            ProtocolMessages.GET_BLOCK_TXS: self.handle_get_block_txs,
+            ProtocolMessages.TRANSACTION: self.handle_transaction,
+            ProtocolMessages.GET_PEER_BLOCK_HASHES: self.handle_get_peer_block_hashes,
+            ProtocolMessages.PEER_BLOCK_HASHES: self.handle_peer_block_hashes,
+            ProtocolMessages.STOP_BLOCK_STREAMING: self.handle_stop_block_streaming,
+            ProtocolMessages.GET_TIPS: self.handle_get_tips,
+            ProtocolMessages.TIPS: self.handle_tips,
+            ProtocolMessages.TIPS_END: self.handle_tips_end,
+            # XXX: overriding ReadyState.handle_error
+            ProtocolMessages.ERROR: self.handle_error,
+            ProtocolMessages.GET_DATA: self.handle_get_data,
+            ProtocolMessages.DATA: self.handle_data,
+            ProtocolMessages.RELAY: self.handle_relay,
+        }
+
+    def handle_error(self, payload: str) -> None:
+        """ Override protocols original handle_error so we can recover a sync in progress.
+        """
+        assert self.protocol.connections is not None
+        sync_checkpoints = self.protocol.connections.sync_checkpoints
+        if sync_checkpoints.is_running and sync_checkpoints.peer_syncing == self:
+            # Oops, we're syncing and we received an error, remove ourselves and let it recover
+            sync_checkpoints.peer_syncing = None
+        # forward message to overloaded handle_error:
+        self.protocol.handle_error(payload)
+
+    def update_synced(self, synced: bool) -> None:
+        self.synced = synced
+
+    def sync_checkpoints_finished(self) -> None:
+        self.state = PeerState.SYNCING_BLOCKS
+
+    @inlineCallbacks
+    def run_sync(self) -> Generator[Any, Any, None]:
```

Comment:
Done.

Reply:
Just checked and the sync will stop running in case of an exception. I feel we should move the current code to a `_run_sync(self)` method and change this one to something like this:

```python
self.is_running = True
try:
    return self._run_sync()
finally:
    self.is_running = False
```

Reply:
Fixed. Thanks!